### ISSUE TITLE
Proposal: Envoy proxy integration

### ISSUE THREAD

**Issue by life1347** [fission/fission#1015]
Envoy proxy is a sidecar proxy widely adopted by companies & projects. It provides rich functionalities such as circuit breaker, rate limit and metric statistics while keeping high performance for service which are quite fit FaaS scenarios for fission. Following is a very rough idea for how fission integrates with envoy.

### Present:
![fission](https://user-images.githubusercontent.com/202578/49431291-1e711180-f7e8-11e8-90f5-ee07127bd07c.png)

Currently, every requests are sent to router and when router received requests it first checks service cache then proxies requests to correspond function pods. It works fine for normal use, however, it also brings some problems list below:

1. Long latency: Since router needs to check service cache every time, it adds extra latency to requests when router is under the heavy workload.
2. Lack of request statistics: To gather requests statistics we need to trace and gather requests information send to function pod which is a huge amount of works for the current codebase.

### After integration:
![fission 1](https://user-images.githubusercontent.com/202578/49431303-25981f80-f7e8-11e8-81dc-3686eeca920f.png)

(**NOTICE**: The architecture graph here is a draft for envoy integration. It may be changed in the future.)

The idea of integration is put envoy as sidecar proxy to router and function pods. As you can see there are two lines in the graph the black one and the red one, they represent different paths for requests to send to function pod.

* Black line (slow path):
    * When a request goes to router pod and if it doesn't match any routing rule in envoy, the request will be sent to router for proxying as the old way.
    * Once the request succeed without error, router will set a routing rule to envoy for fast path.  

* Red line (fast path):
    * When requests for the same function arrive router pod, envoy will send them to function pods directly without asking router.

In this way, we can improve router performance while keeping codebase clean.
Also, envoy exposes statistics API for multiple data collectors like prometheus and datadog, users can visualize statistics with tools get used to.

The idea sounds great, there are problems/questions we need to investigate first:
1. Will it conflict with Istio if adding extra envoy proxy to fission?
2. Though envoy supports dynamic configuration & hot restart, what will happen if router refreshes envoy config frequently?
3. The extra resource usage
4. .etc

Since it takes some time to do some experiments, I separate tasks into couple stages:
- [ ] Play around envoy to see what API and functionality we can use
- [ ] Do some experiments
- [ ] MVP for integration
- [ ] Test with Istio (Also fix istio problem: #766 )
- [ ] Formal design spec & architecture graph.
- [ ] TBD

**Comment by soamvasani** [comment]
Thanks for writing this up @life1347, the diagrams are particularly helpful.

I'm also interested in the pros/cons of having envoy in the router but not the function pod.  Because that way function pods can be lighter weight, and we gain a lot of the benefits of envoy simply from having it as a sort of fission router fast path.

**Comment by ajbouh** [comment]
In case it's helpful for planning...

Ambassador initially used envoy's restart logic to load new config. This meant that every time there was a new service to route to, envoy needed to be *restarted*. This drained and closed existing connections which meant that it was not a good fit for large traffic loads.

To work around this, they're rewriting their config logic to use envoy's ADS interface: https://github.com/datawire/ambassador/issues/440. This rewrite is almost complete: https://blog.getambassador.io/ambassador-0-50-ea7-available-70631a836875.

None of this is meant to say that ADS *must* be used, or that Ambassador's decisions were right or wrong. Just calling out a similar journey taken by another project in the Kubernetes ecosystem.

I agree with @soamvasani that envoy as a fast path in router might be worth it even if envoy isn't used in function pods.

**Comment by life1347** [comment]
@soamvasani 
I agree to put envoy as sidecar proxy in the router pod for fast path purpose, however, from the [tracing doc ](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/tracing) looks like we still need to add envoy to function pod as well to gather some useful information. (maybe I'm wrong here) I will do some experiments to see the pros/cons for with/without envoy in function pod.

@ajbouh 
Yes, I do think we need to implement xDS API since hot restart is a little bit heavy for fission use cases 
(or maybe a hybrid one?) and there are some tests need to be took place first like how long the envoy takes to reflect the changes in router xDS API. Anyway, thanks for providing such useful information before we start our own journey. It really helps a lot.  👍

**Comment by ajbouh** [comment]
The tracing support is something that's easy to include in your function implementation using various opentracing client libraries. It's best to submit this from within your functions even if you have an envoy sidecar because then you can include fine-grained events that are not visible outside your function's container.

**Comment by soamvasani** [comment]
It's worth noting that we'll get plenty of tracing data from just the router.  All function calls go through the router (and in this design the router pod's envoy) so we'll get function span data from just having one envoy.  (Though it will be less accurate since it'll have network latencies mixed into the function span run time.)

Also, we already co-exist nicely with istio and we can continue that.  So if users want envoy in every pod they can just use a service mesh, right?

**Comment by life1347** [comment]
> So if users want envoy in every pod they can just use a service mesh

Make sense to me, didn't think of this before. Ok, I will head over this direction first.

**Comment by sanketsudake** [comment]
Fission now integrates Envoy at the routing layer via the **Gateway API route provider** (#3478), which uses Envoy Gateway as the data path (the legacy Ingress provider is deprecated in favor of it).

This brings Envoy's capabilities into the request path as this proposal envisioned back in 2018, via the now-standard Gateway API rather than a bespoke integration.
Closing as superseded by the Gateway API provider — thanks for the original proposal.
